pupil_src/shared_modules/annotations.py

```diff
@@ -7,239 +7,329 @@
     """
+
+    class _AnnotationInfo:
+        # no namedtuple because we cannot set the menu button at creation
+        # (this button references the _AnnotationInfo for deletion)
+        def __init__(self, definition, button_quickbar=None, button_menu=None):
+            self.definition = definition
+            self.button_quickbar = button_quickbar
+            self.button_menu = button_menu
+
     icon_chr = chr(0xe866)
-    icon_font = 'pupil_icons'
+    icon_font = "pupil_icons"
 
-    def __init__(self, g_pool, annotation_definitions=(('My annotation', 'E'),)):
+    def __init__(self, g_pool, annotation_definitions=None):
         super().__init__(g_pool)
         self.menu = None
-        self.sub_menu = None
-        self.buttons = []
+        self._annotation_list_menu = None
 
-        self.annotation_definitions = list(annotation_definitions)
+        if annotation_definitions is None:
+            annotation_definitions = [["My annotation", "E"]]
+        self._initial_annotation_definitions = annotation_definitions
+        self._annotation_infos = []
 
-        self.new_annotation_name = 'new annotation name'
-        self.new_annotation_hotkey = 'e'
+        self._new_annotation_name = "new annotation name"
+        self._new_annotation_hotkey = "E"
 
-        self.current_frame = -1
+    def get_init_dict(self):
+        annotation_definitions = [info.definition for info in self._annotation_infos]
+        return {"annotation_definitions": annotation_definitions}
 
     def init_ui(self):
         self.add_menu()
-        self.menu.label = 'Annotations'
-        self.menu.append(ui.Text_Input('new_annotation_name',self))
-        self.menu.append(ui.Text_Input('new_annotation_hotkey',self))
-        self.menu.append(ui.Button('add annotation type',self.append_annotation))
-        self.sub_menu = ui.Growing_Menu('Events - click to remove')
-        self.menu.append(self.sub_menu)
-        self.update_buttons()
-
-    def update_buttons(self):
-        for b in self.buttons:
-            self.g_pool.quickbar.remove(b)
-            self.sub_menu.elements[:] = []
-        self.buttons = []
-
-        for e_name, hotkey in self.annotation_definitions:
-            def make_fire(e_name, hotkey):
-                return lambda _ : self.fire_annotation(e_name)
-
-            def make_remove(e_name, hotkey):
-                return lambda: self.remove_annotation((e_name, hotkey))
-
-            thumb = ui.Thumb(e_name, setter=make_fire(e_name, hotkey),
-                             getter=lambda: False, label=hotkey, hotkey=hotkey)
-            self.buttons.append(thumb)
-            self.g_pool.quickbar.append(thumb)
-            self.sub_menu.append(ui.Button(e_name+" <"+hotkey+">", make_remove(e_name, hotkey)))
+        self.customize_menu()
+        self.menu.append(ui.Text_Input("_new_annotation_name", self, label="New Name"))
+        self.menu.append(ui.Text_Input("_new_annotation_hotkey", self, label="New Hotkey"))
```

Comment:
Line over 88 characters long

```suggestion
        self.menu.append(
            ui.Text_Input("_new_annotation_hotkey", self, label="New Hotkey")
        )
```